### ISSUE TITLE
spread syntax

### ISSUE THREAD

**Issue by mccolljr** [d5/tengo#190]
# Spread Syntax

Now that the language has variadic functions, it would make sense to add something like Go's spread operator (`...`) to allow using the contents of arrays as variadic arguments.

However, Tengo uses a stack-based vm. Implementing a spread-like operator would create a re-usable mechanism for "exploding" arrays onto the stack. We can take advantage of this by allowing "exploding" in array literals and at other positions in a function call besides the last.

I've got most of an implementation complete in the explosions branch of my fork.
For syntax and uses, see below.

#### Array Literals
insert/delete:
```go
x := [1, 2, 3, 4]
x = [x[:2]..., "hi", x[2:]...] // [1, 2, "hi", 4, 5]
x = [x[:2]..., x[3:]...] // [1, 2, 4, 5]
```
concatenate:
```go
a := [1, 2]
b := [3, 4]
c := [5, 6]
d := [a..., b..., c...] // [1, 2, 3, 4, 5, 6]
```
mix & match:
```go
a := [2, 3]
b := [5, 6]
c := [1, a..., 4, b...] // [1, 2, 3, 4, 5, 6]
```

explode custom `Iterable` implementations:
```go
iter := make_go_channel_iterator()
list := [iter...] // [ (all values from the iterator) ]
```

explode any value of an explodable type (including literals):
```go
x := [[1, 2]..., 3, [4, 5, 6]...] // [1, 2, 3, 4, 5, 6]
```

#### Function Calls

our example funcs:
```go
variadic_fn := func(a, b, c, ...args) { /* ... */ }
normal_fn := func(a, b, c, d) { /* ... */ } 
```

variadic function calls:
```go
variadic_fn(1, 2, 3) // a=1 b=2 c=3 args=[]
variadic_fn(1, 2, 3, [4, 5]...) // a=1 b=2 c=3 args=[4, 5]
variadic_fn(1, [2, 3]..., [4, 5]..., 6) // a=1 b=2 c=3 args=[4, 5, 6]
```

normal function calls:
```go
normal_fn(1, 2, 3) // illegal, missing arguments
normal_fn(1, 2, 3, [4, 5]...) // illegal, total # of arguments > # of function parameters
normal_fn(1, [2, 3]..., [4]...) // a=1 b=2 c=3 d=4
normal_fn([1, 2, 3, 4]...) // a=1, b=2, c=3, d=4
```

example:
```go
apply := func(fn, ...applied) {
    return func(...args) {
        return fn(applied..., args...)
    }
}

apply(normal_fn, 1, 2)(3, 4) // a=1 b=2 c=3 d=4
```

**Comment by earncef** [comment]
I'm not sure about map explosions for two reasons:
- Maps don't guarantee the order of elements
- I'm not entirely sure something like `[["a", 1], ["b", 2], ["c", 3]]` will actually add anything useful that you can't do with map iterations.

It's the same reason I'm not sure this should be applicable for any `Iterable` implementations other than arrays/immutable arrays unless the `Object` can be checked for a guarantee on the order of elements.

**Comment by mccolljr** [comment]
I updated the proposal to acknowledge lack of map ordering guarantees. I can see where you're coming from with not including that, but I think as long as we document the fact that we only explode the values and that ordering guarantees are implementation-dependent, including custom iterators would be useful.

You could collect an iterator, or collect several iterators in sequence in one statement:
```go
collected := [custom_iter...]
all_collected := [ter_1..., iter_2..., iter_3...]
```

Having that shorthand syntax is nice, but IMO the power really comes from implementing custom explodable types.

You can pass long sequences as compact values that can be exploded when needed:
```go
lazy100 = lazy_range(0,100) // passed around as go value of [2]int
lazy100_by_2 = lazy_range(0,100,2) // [3]int
lazy100_by_2_offset_1 = lazy_range(0,100,2,1) // [4]int
/* ... a while later ... */
evens := [lazy100_by_2...] // [0, 2, 4, ..., 98]
odds := [lazy100_by_2_offset_1...] // [1, 3, 5, ..., 99]
```

You can explode custom types as function arguments:
```go
origin := tuple(0, 0)
set := func(x, y, val) { /* ... */ }
setOrigin := func(val) { set(origin..., val) }
```

**Comment by earncef** [comment]
I’m not against custom explodable types per se but I am not convinced about explodable maps or objects that can’t guarantee order of elements. While documentation might help, it is very non-intuitive and is likely to create confusion, given the kinds of cases where variables can be exploded.

- exploding a map into an array is nonintuitive as array guarantees order and map doesn’t
- exploding into function calls assumes ordering variadic or not (variadic param is an array) which is again a problem with maps

As for the case of merging iterators, there are several ways it could be achieved without the confusion caused by an explosion. For instance, it could be just as easily be done by implementing BinaryOp() to use ‘+’ to merge objects into a new copy. Or a function to do something similar.

**Comment by earncef** [comment]
An alternative would be two operators, one that works on ordered elements and another that works on all but without order guarantees.

**Comment by earncef** [comment]
Another problem with the map implementation as you’ve described it is that it’s inconsistent with all the other iterable implementations. All the others only explode values, map explodes an arrays of key and value. 

Maps are not guarantees to be the only Objects with non-int keys. This inconsistency needs to be dealt with.

**Comment by mccolljr** [comment]
I removed maps from the proposal - it's potentially confusing functionality without a compelling use case. I see your point there. We can even go so far as to prevent maps from being exploded with a runtime error - it's better to limit it to arrays and types where the responsibility for iteration order is up to the user.

For custom explodable types, thoughts on adding a new interface to the `objects` package that custom objects can implement? This would eliminate confusion over how an `Iterable` explodes by requiring that the object returns values specifically meant to be used in an explosion.
```go
type Explodable interface {
    Explode() (values []Object, err error)
}
```

**Comment by earncef** [comment]
I'm good with the `Explodable` interface.

**Comment by jcw** [comment]
FWIW, I'd vote to _avoid_ a new interface for spreads, for lack of a clear need. While array spreads will be very convenient, everything else can be turned into an array before getting used with `...` - I think.

(When in any doubt, please consider _not_ adding a new feature yet. It's always very hard to get the genie back into the bottle. At least that's the lesson I always take away from the Go language developers ...)

**Comment by mccolljr** [comment]
implemented by #203, with the caveat that only the last position in a call may contain a spread expression.
